server/src/compression.rs

```diff
@@ -0,0 +1,473 @@
+    /// First timestamp of each compressed segment in the batch.
+    start_times: TimestampBuilder,
+    /// Last timestamp of each compressed segment in the batch.
+    end_times: TimestampBuilder,
+    /// Data required in addition to `min_value` and `max_value` to reconstruct
+    /// the values of each compressed segment in the batch within an error
+    /// bound.
+    values: BinaryBuilder,
+    /// Minimum value of each compressed segment in the batch.
+    min_values: ValueBuilder,
+    /// Maximum value of each compressed segment in the batch.
+    max_values: ValueBuilder,
+    /// Actual error of each compressed segment in the batch.
+    error: Float32Builder,
+}
+
+impl CompressedRecordBatchBuilder {
+    fn new(capacity: usize) -> Self {
+        Self {
+            model_type_ids: UInt8Builder::new(capacity),
+            timestamps: BinaryBuilder::new(capacity),
+            start_times: TimestampBuilder::new(capacity),
+            end_times: TimestampBuilder::new(capacity),
+            values: BinaryBuilder::new(capacity),
+            min_values: ValueBuilder::new(capacity),
+            max_values: ValueBuilder::new(capacity),
+            error: Float32Builder::new(capacity),
+        }
+    }
+
+    /// Append a compressed segment to the builder.
+    fn append_compressed_segment(
+        &mut self,
+        model_type_id: u8,
+        timestamps: &[u8],
+        start_time: Timestamp,
+        end_time: Timestamp,
+        values: &[u8],
+        min_value: Value,
+        max_value: Value,
+        error: f32,
+    ) {
+        // unwrap() is used as append_value() never returns Error for
+        // PrimitiveBuilder.
+        self.model_type_ids.append_value(model_type_id).unwrap();
+        self.timestamps.append_value(timestamps).unwrap();
+        self.start_times.append_value(start_time).unwrap();
+        self.end_times.append_value(end_time).unwrap();
+        self.values.append_value(values).unwrap();
+        self.min_values.append_value(min_value).unwrap();
+        self.max_values.append_value(max_value).unwrap();
+        self.error.append_value(error).unwrap();
+    }
+
+    /// Return `RecordBatch` of compressed segments and consume the builder.
+    fn finnish(mut self) -> RecordBatch {
+        RecordBatch::try_new(
+            Arc::new(StorageEngine::get_compressed_segment_schema()),
+            vec![
+                Arc::new(self.model_type_ids.finish()),
+                Arc::new(self.timestamps.finish()),
+                Arc::new(self.start_times.finish()),
+                Arc::new(self.end_times.finish()),
+                Arc::new(self.values.finish()),
+                Arc::new(self.min_values.finish()),
+                Arc::new(self.max_values.finish()),
+                Arc::new(self.error.finish()),
+            ],
+        )
+        .unwrap()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use datafusion::arrow::array::UInt8Array;
+
+    // Tests compress.
```

Comment:
Maybe `Tests for try_compress().` instead.

Reply:
Fixed.

Reply:
Good catch, fixed.